packages/docregistry/src/default.ts

```diff
@@ -387,6 +393,11 @@ export abstract class ABCWidgetFactory<
    */
   protected abstract createNewWidget(context: DocumentRegistry.IContext<U>): T;
 
+  /**
+   * toolbar items to be added when createNew
+   */
+  toolbarItems?: DocumentRegistry.IToolbarItem[];
```

Comment:
I think this should always be available, and be an empty list by default.

Reply:
What I mean is that it should be optional in `IOptions` but concrete on `ABCWidgetFactory`, where we set it to `options.toolbarItems || []` in the constructor.

Reply:
I tried to put the code `this.toolbarItems = options.toolbarItems || []` in `ABCWidgetFactory.constructor`, but then later in `NotebookWidgetFactory.createNewWidget` the below code:

```js
if (!this.toolbarItems) {
      this.toolbarItems = ToolbarItems.getDefaultItems(widget);
}
```

cannot work, since the `options.toolbarItems` is lost, if `this.toolbarItems` is `[]`, we don't know whether it's user intended or just empty to want to fallback to default toolbar.

I agree with you that it's better to have a solid type for `this.toolbarItems`, the `null` is evil.

One way I can image is to save another state that whether fallback to default toolbar? Or just provide a way to let user populate default toolbar?

Reply:
Hmm, okay, I see your point.  Maybe it should be a function that we pass in as the option instead: `(widget: T) => IToolbarItems`.  The `NotebookWidgetFactory` would default to `ToolbarItems.getDefaultItems`.

Reply:
I see, it's can be done but need to pass down `<T>` as `DocumentRegistry.IWidgetFactoryOptions<T>`, many existing code rely on it will be changed.

Maybe another option is to define a function `createToolbarItems` in `ABCWidgetFactory`?

```js
createToolbarItems(widget: T): DocumentRegistry.IToolbarItem[] {
    return [];
}
```

Then

```js
createNew(context: DocumentRegistry.IContext<U>): T {
    let widget = this.createNewWidget(context);
    let toolbarItems = this.createToolbarItems(widget);
    toolbarItems.forEach(({ name, widget: item }) => {
      widget.toolbar.addItem(name, item);
    });
    this._widgetCreated.emit(widget);
    return widget;
}
```

The `NotebookWidgetFactory` need to override  `createToolbarItems` method to populate default items.

Reply:
OK, let's looking back to the root of the problem:

`ABCWidgetFactory.toolbarItems: DocumentRegistry.IToolbarItem[];` can be set to `null`?

This is fine if there's no plan to enable `--strictNullChecks` in typescript. Otherwise I think this is fine now. What do you think?

If override `IOptions.toolbarItems`, the same thing here, if enabed `strictNullChecks`, that value cannot be stored if it's `null`. If use `IOptions.createToolbarItems(widget)` then it's may rely on `<T>`.

Reply:
I double check the `createToolbarItems`  class method, found it's redundant, since user can did it when override `createNewWidget`,  and set `this.toolbarItems`, so please ignore the `createToolbarItems` proposal.

The root problem still is whether the `ABCWidgetFactory.toolbarItems` can be null, and how to deal with it.

Reply:
I don't think it needs to be a public attribute.  I think it should be `private _toolbarItems: DocumentRegistry.IToolbarItem[] | null`.    Hmm, what if we added them in `createNew()` if they were given as an open, and then in `createNewWidget()` we can do whatever we want at that point.  In the case of a notebook, if the toolbar has no items, we add our defaults.

Reply:
You could call it `_defaultToolbarItems`, I think that would be clearer.

Reply:
The `NotebookWidgetFactory.IOptions` has `toolbarItems` too, need to store this then in `createNewWidget` to decide whether to populate default items, maybe in `ABCWidgetFactory`

```js
protected _defaultToolbarItems: DocumentRegistry.IToolbarItem[] | undefined;
```
is better?

Reply:
How about `private _overrideToolbarItems: DocumentRegistry.IToolbarItem[] | undefined` and `protected defaultToolbarItems: DocumentRegistry.IToolbarItem[]`?

Reply:
sorry I don't get your point... I think the use case can make me more clear:

1. 
```js
new NotebookWidgetFactory({name: 'notebook', fileTypes: ['notebook'], toolbarItems: [A, B]})
// 'default-toolbar.ts':  getDefault: [X, Y, Z]
```

What will be the value of `_overrideToolbarItems ` and `defaultToolbarItems `?

2.

```js
new NotebookWidgetFactory({name: 'notebook', fileTypes: ['notebook'], toolbarItems: undefined})
```
This one should set `widget.toolbar` to  [X, Y, Z]
What will be the value of `_overrideToolbarItems ` and `defaultToolbarItems `?

Reply:
I've pushed new commit to reflect the latest discussion.

Reply:
I think we're over-complicating this.  Looking back at your original ask, all we wanted was the ability to override the default via `IOptions`.  In light of that, I propose:

1.  `IOptions.overrideToolbarItems?: IToolbarItem`
2. `protected `overrideToolbarItems: IToolbarItem[] | undefined`
3.  In `createWidget()`, add `_overrideToolbarItems` if it is defined.
4.  In `NotebookWidgetFactory.createNewWidget()`, if `overrideToolbarItems` is undefined, use `ToolbarItems.getDefaultItems()` to populate.